File: python_hft/models/engine.py
import threading
import time
from datetime import datetime, timedelta
from collections import defaultdict, deque
import queue

from models.order import Order, OrderSide, OrderStatus
from models.orderbook import OrderBook

class TradingEngine:
    """
    Main trading engine that handles order matching and execution
    """
    
    def __init__(self):
        """Initialize the trading engine"""
        self.orderbooks = {}  # symbol -> OrderBook
        self.active_orders = {}  # order_id -> order
        self.traders = {}  # trader_id -> trader reference
        
        # Trade execution tracking
        self.trade_history = deque(maxlen=10000)
        self.executions_today = 0
        
        # Performance metrics
        self.start_time = datetime.now()
        self.total_trades = 0
        self.total_volume = 0
        self.latency_measurements = deque(maxlen=1000)
        
        # Threading
        self.is_running = False
        self.order_queue = queue.Queue()
        self.execution_thread = None
        self.stats_lock = threading.Lock()
        self.orders_lock = threading.Lock()
        
        # Order processing statistics (optimized for HFT)
        self.orders_per_second = 0
        self.last_stats_update = time.time()
        self.orders_processed_since_last_update = 0
        self.batch_size = 100  # Process orders in batches for better performance
    
    def start(self):
        """Start the trading engine"""
        if not self.is_running:
            self.is_running = True
            self.execution_thread = threading.Thread(target=self._execution_loop, daemon=True)
            self.execution_thread.start()
    
    def stop(self):
        """Stop the trading engine"""
        self.is_running = False
        if self.execution_thread and self.execution_thread.is_alive():
            self.execution_thread.join(timeout=2.0)
    
    def register_trader(self, trader):
        """Register a trader with the engine"""
        self.traders[trader.trader_id] = trader
    
    def get_orderbook(self, symbol):
        """Get or create order book for a symbol"""
        if symbol not in self.orderbooks:
            self.orderbooks[symbol] = OrderBook(symbol)
        return self.orderbooks[symbol]
    
    def submit_order(self, order):
        """Submit an order for processing"""
        order_submit_time = time.time()
        order.submit_time = order_submit_time
        
        # Add to queue for processing
        self.order_queue.put(order)
    
    def _execution_loop(self):
        """Main execution loop that processes orders (optimized for HFT)"""
        while self.is_running:
            try:
                # Process orders in batches for better performance
                orders_to_process = []
                
                # Collect batch of orders with minimal blocking
                start_time = time.time()
                while len(orders_to_process) < self.batch_size and (time.time() - start_time) < 0.001:
                    try:
                        order = self.order_queue.get_nowait()
                        orders_to_process.append(order)
                    except queue.Empty:
                        break
                
                # Process the batch
                for order in orders_to_process:
                    self._process_order(order)
                    self.order_queue.task_done()
                
                # If no orders, small sleep to prevent busy waiting
                if not orders_to_process:
                    time.sleep(0.0001)  # 0.1ms sleep for HFT performance
                    
            except Exception as e:
                print(f"Error in execution loop: {e}")
                time.sleep(0.001)  # Brief pause on error
    
    def _process_order(self, order):
        """Process a single order"""
        process_start = time.time()
        
        with self.orders_lock:
            # Add to active orders
            self.active_orders[order.order_id] = order
            
            # Get order book
            orderbook = self.get_orderbook(order.symbol)
            
            # Try to match the order
            self._match_order(order, orderbook)
            
            # If order still has quantity, add to book
            if order.is_active() and order.quantity > 0:
                orderbook.add_order(order)
            else:
                # Remove from active orders if completely filled or cancelled
                if order.order_id in self.active_orders:
                    del self.active_orders[order.order_id]
        
        # Record processing latency
        process_end = time.time()
        latency_ms = (process_end - process_start) * 1000
        
        if hasattr(order, 'submit_time'):
            total_latency_ms = (process_end - order.submit_time) * 1000
            self.latency_measurements.append(total_latency_ms)
        
        # Update statistics
        self._update_processing_stats()
    
    def _match_order(self, incoming_order, orderbook):
        """Match an incoming order against the order book"""
        if incoming_order.side == OrderSide.BUY:
            # Match buy order against asks (sell orders)
            self._match_buy_order(incoming_order, orderbook)
        else:
            # Match sell order against bids (buy orders)
            self._match_sell_order(incoming_order, orderbook)
    
    def _match_buy_order(self, buy_order, orderbook):
        """Match a buy order against asks"""
        while buy_order.quantity > 0 and buy_order.is_active():
            best_ask = orderbook.get_best_ask()
            
            if not best_ask or best_ask.price > buy_order.price:
                break  # No more matching orders
            
            # Execute trade
            trade_quantity = min(buy_order.quantity, best_ask.quantity)
            trade_price = best_ask.price  # Price-time priority: use ask price
            
            self._execute_trade(buy_order, best_ask, trade_quantity, trade_price, orderbook)
            
            # Remove ask order if completely filled
            if best_ask.quantity == 0:
                orderbook.remove_order(best_ask.order_id, OrderSide.SELL)
                if best_ask.order_id in self.active_orders:
                    del self.active_orders[best_ask.order_id]
    
    def _match_sell_order(self, sell_order, orderbook):
        """Match a sell order against bids"""
        while sell_order.quantity > 0 and sell_order.is_active():
            best_bid = orderbook.get_best_bid()
            
            if not best_bid or best_bid.price < sell_order.price:
                break  # No more matching orders
            
            # Execute trade
            trade_quantity = min(sell_order.quantity, best_bid.quantity)
            trade_price = best_bid.price  # Price-time priority: use bid price
            
            self._execute_trade(best_bid, sell_order, trade_quantity, trade_price, orderbook)
            
            # Remove bid order if completely filled
            if best_bid.quantity == 0:
                orderbook.remove_order(best_bid.order_id, OrderSide.BUY)
                if best_bid.order_id in self.active_orders:
                    del self.active_orders[best_bid.order_id]
    
    def _execute_trade(self, buy_order, sell_order, quantity, price, orderbook):
        """Execute a trade between two orders"""
        trade_time = datetime.now()
        
        # Fill both orders
        buy_order.fill(quantity, price)
        sell_order.fill(quantity, price)
        
        # Create trade record
        trade = {
            'trade_id': f"{self.total_trades + 1:06d}",
            'timestamp': trade_time,
            'symbol': buy_order.symbol,
            'quantity': quantity,
            'price': price,
            'buyer_id': buy_order.trader_id,
            'seller_id': sell_order.trader_id,
            'buy_order_id': buy_order.order_id,
            'sell_order_id': sell_order.order_id,
            'side': 'BUY'  # From the perspective of the aggressive order
        }
        
        # Add to trade history
        with self.stats_lock:
            self.trade_history.append(trade)
            self.total_trades += 1
            self.total_volume += quantity
        
        # Add to order book trade history
        orderbook.add_trade(trade)
        
        # Notify traders of fills
        self._notify_trader_fill(buy_order, quantity, price)
        self._notify_trader_fill(sell_order, quantity, price)
    
    def _notify_trader_fill(self, order, quantity, price):
        """Notify a trader that their order was filled"""
        trader_id = order.trader_id
        if trader_id in self.traders:
            try:
                trader = self.traders[trader_id]
                trader.on_order_filled(order, quantity, price)
            except Exception as e:
                print(f"Error notifying trader {trader_id}: {e}")
    
    def _update_processing_stats(self):
        """Update processing statistics"""
        current_time = time.time()
        self.orders_processed_since_last_update += 1
        
        # Update stats every second
        if current_time - self.last_stats_update >= 1.0:
            self.orders_per_second = self.orders_processed_since_last_update
            self.orders_processed_since_last_update = 0
            self.last_stats_update = current_time
    
    def cancel_order(self, order_id):
        """Cancel an order"""
        with self.orders_lock:
            if order_id in self.active_orders:
                order = self.active_orders[order_id]
                order.cancel()
                
                # Remove from order book
                orderbook = self.get_orderbook(order.symbol)
                orderbook.remove_order(order_id, order.side)
                
                # Remove from active orders
                del self.active_orders[order_id]
                return True
        return False
    
    def get_recent_trades(self, count=20):
        """Get recent trades across all symbols"""
        with self.stats_lock:
            return list(self.trade_history)[-count:] if count > 0 else list(self.trade_history)
    
    def get_recent_trades_for_symbol(self, symbol, count=10):
        """Get recent trades for a specific symbol"""
        with self.stats_lock:
            symbol_trades = [trade for trade in self.trade_history if trade['symbol'] == symbol]
            return symbol_trades[-count:] if count > 0 else symbol_trades
    
    def get_all_trades(self):
        """Get all trades for export"""
        with self.stats_lock:
            return list(self.trade_history)
    
    def get_performance_stats(self):
        """Get engine performance statistics"""
        with self.stats_lock:
            current_time = datetime.now()
            runtime_seconds = (current_time - self.start_time).total_seconds()
            
            # Calculate trades per second
            trades_per_second = self.total_trades / max(1, runtime_seconds)
            
            # Calculate average latency
            avg_latency_ms = 0
            if self.latency_measurements:
                avg_latency_ms = sum(self.latency_measurements) / len(self.latency_measurements)
            
            # Count active orders
            active_orders_count = len(self.active_orders)
            
            return {
                'total_trades': self.total_trades,
                'total_volume': self.total_volume,
                'trades_per_second': trades_per_second,
                'orders_per_second': self.orders_per_second,
                'avg_latency_ms': avg_latency_ms,
                'active_orders': active_orders_count,
                'runtime_seconds': runtime_seconds,
                'symbols_active': len(self.orderbooks)
            }
    
    def get_market_summary(self):
        """Get summary of all markets"""
        summary = {}
        
        for symbol, orderbook in self.orderbooks.items():
            stats = orderbook.get_statistics()
            recent_trades = orderbook.get_recent_trades(5)
            
            # Calculate volume-weighted average price (VWAP) for recent trades
            vwap = 0
            if recent_trades:
                total_value = sum(trade['price'] * trade['quantity'] for trade in recent_trades)
                total_volume = sum(trade['quantity'] for trade in recent_trades)
                if total_volume > 0:
                    vwap = total_value / total_volume
            
            summary[symbol] = {
                'best_bid': stats['mid_price'],
                'best_ask': stats['mid_price'],
                'spread': stats['spread'],
                'mid_price': stats['mid_price'],
                'vwap': vwap,
                'volume': stats['total_bid_volume'] + stats['total_ask_volume'],
                'trade_count': len(recent_trades)
            }
        
        return summary
    
    def get_trader_orders(self, trader_id):
        """Get all active orders for a trader"""
        with self.orders_lock:
            return [order for order in self.active_orders.values() 
                    if order.trader_id == trader_id]
    
    def get_symbol_statistics(self, symbol):
        """Get detailed statistics for a symbol"""
        if symbol not in self.orderbooks:
            return None
        
        orderbook = self.orderbooks[symbol]
        recent_trades = self.get_recent_trades_for_symbol(symbol, 100)
        
        # Calculate price statistics
        if recent_trades:
            prices = [trade['price'] for trade in recent_trades]
            high_price = max(prices)
            low_price = min(prices)
            last_price = prices[-1] if prices else 0
            
            # Calculate VWAP
            total_value = sum(trade['price'] * trade['quantity'] for trade in recent_trades)
            total_volume = sum(trade['quantity'] for trade in recent_trades)
            vwap = total_value / total_volume if total_volume > 0 else 0
        else:
            high_price = low_price = last_price = vwap = 0
        
        return {
            'symbol': symbol,
            'last_price': last_price,
            'high_price': high_price,
            'low_price': low_price,
            'vwap': vwap,
            'total_volume': sum(trade['quantity'] for trade in recent_trades),
            'trade_count': len(recent_trades),
            'orderbook_stats': orderbook.get_statistics()
        }

File: cpp_hft/src/engine.cpp
#include "engine.h"
#include <QMutexLocker>
#include <QDebug>
#include <QUuid>
#include <algorithm>

TradingEngine::TradingEngine(QObject* parent)
    : QObject(parent)
    , totalTrades(0)
    , totalVolume(0)
    , running(false)
    , processingTimer(new QTimer(this))
    , statsTimer(new QTimer(this))
    , ordersPerSecond(0)
    , ordersProcessedSinceLastUpdate(0)
{
    // Connect timers
    connect(processingTimer, &QTimer::timeout, this, &TradingEngine::processOrders);
    connect(statsTimer, &QTimer::timeout, this, &TradingEngine::updateStats);
    
    // Set timer intervals
    processingTimer->setSingleShot(false);
    processingTimer->setInterval(1); // 1ms for high frequency processing
    
    statsTimer->setSingleShot(false);
    statsTimer->setInterval(1000); // 1 second for stats updates
}

TradingEngine::~TradingEngine() {
    stop();
}

void TradingEngine::start() {
    if (!running) {
        running = true;
        startTime.start();
        lastStatsUpdate.start();
        
        processingTimer->start();
        statsTimer->start();
        
        qDebug() << "Trading engine started";
    }
}

void TradingEngine::stop() {
    if (running) {
        running = false;
        processingTimer->stop();
        statsTimer->stop();
        
        qDebug() << "Trading engine stopped";
    }
}

void TradingEngine::registerTrader(Trader* trader) {
    if (trader) {
        traders[trader->getTraderId()] = trader;
        
        // Connect trader signals
        connect(this, &TradingEngine::orderFilled, 
                trader, &Trader::onOrderFilled);
    }
}

std::shared_ptr<OrderBook> TradingEngine::getOrderBook(const QString& symbol) {
    if (!orderBooks.contains(symbol)) {
        orderBooks[symbol] = std::make_shared<OrderBook>(symbol);
    }
    return orderBooks[symbol];
}

void TradingEngine::submitOrder(OrderPtr order) {
    QMutexLocker locker(&orderQueueMutex);
    orderQueue.enqueue(order);
}

bool TradingEngine::cancelOrder(const QString& orderId) {
    QMutexLocker locker(&orderQueueMutex);
    
    if (activeOrders.contains(orderId)) {
        OrderPtr order = activeOrders[orderId];
        order->cancel();
        
        // Remove from order book
        auto orderBook = getOrderBook(order->getSymbol());
        orderBook->removeOrder(orderId, order->getSide());
        
        // Remove from active orders
        activeOrders.remove(orderId);
        return true;
    }
    return false;
}

QVector<Trade> TradingEngine::getRecentTrades(int count) const {
    QMutexLocker locker(&tradeHistoryMutex);
    
    if (count <= 0 || tradeHistory.isEmpty()) {
        return tradeHistory;
    }
    
    int startIndex = qMax(0, tradeHistory.size() - count);
    return tradeHistory.mid(startIndex);
}

QVector<Trade> TradingEngine::getRecentTradesForSymbol(const QString& symbol, int count) const {
    QMutexLocker locker(&tradeHistoryMutex);
    
    QVector<Trade> symbolTrades;
    for (const auto& trade : tradeHistory) {
        if (trade.symbol == symbol) {
            symbolTrades.append(trade);
        }
    }
    
    if (count <= 0 || symbolTrades.isEmpty()) {
        return symbolTrades;
    }
    
    int startIndex = qMax(0, symbolTrades.size() - count);
    return symbolTrades.mid(startIndex);
}

QVector<Trade> TradingEngine::getAllTrades() const {
    QMutexLocker locker(&tradeHistoryMutex);
    return tradeHistory;
}

QMap<QString, QVariant> TradingEngine::getPerformanceStats() const {
    QMutexLocker locker(&statsMutex);
    
    QMap<QString, QVariant> stats;
    
    qint64 runtimeMs = startTime.elapsed();
    double runtimeSeconds = runtimeMs / 1000.0;
    
    // Calculate trades per second
    double tradesPerSecond = totalTrades / qMax(1.0, runtimeSeconds);
    
    // Calculate average latency
    double avgLatencyMs = 0.0;
    if (!latencyMeasurements.isEmpty()) {
        qint64 totalLatency = 0;
        for (qint64 latency : latencyMeasurements) {
            totalLatency += latency;
        }
        avgLatencyMs = totalLatency / static_cast<double>(latencyMeasurements.size());
    }
    
    // Count active orders
    int activeOrdersCount = activeOrders.size();
    
    stats["total_trades"] = totalTrades;
    stats["total_volume"] = totalVolume;
    stats["trades_per_second"] = tradesPerSecond;
    stats["orders_per_second"] = ordersPerSecond;
    stats["avg_latency_ms"] = avgLatencyMs;
    stats["active_orders"] = activeOrdersCount;
    stats["runtime_seconds"] = runtimeSeconds;
    stats["symbols_active"] = orderBooks.size();
    
    return stats;
}

QMap<QString, QVariant> TradingEngine::getMarketSummary() const {
    QMap<QString, QVariant> summary;
    
    for (auto it = orderBooks.begin(); it != orderBooks.end(); ++it) {
        const QString& symbol = it.key();
        auto orderBook = it.value();
        
        auto recentTrades = orderBook->getRecentTrades(5);
        
        // Calculate volume-weighted average price (VWAP) for recent trades
        double vwap = 0.0;
        if (!recentTrades.isEmpty()) {
            double totalValue = 0.0;
            int totalVolume = 0;
            
            for (const auto& trade : recentTrades) {
                totalValue += trade.price * trade.quantity;
                totalVolume += trade.quantity;
            }
            
            if (totalVolume > 0) {
                vwap = totalValue / totalVolume;
            }
        }
        
        QMap<QString, QVariant> symbolSummary;
        symbolSummary["best_bid"] = orderBook->getBestBidPrice();
        symbolSummary["best_ask"] = orderBook->getBestAskPrice();
        symbolSummary["spread"] = orderBook->getSpread();
        symbolSummary["mid_price"] = orderBook->getMidPrice();
        symbolSummary["vwap"] = vwap;
        symbolSummary["trade_count"] = recentTrades.size();
        
        summary[symbol] = symbolSummary;
    }
    
    return summary;
}

QVector<OrderPtr> TradingEngine::getTraderOrders(const QString& traderId) const {
    QMutexLocker locker(&orderQueueMutex);
    
    QVector<OrderPtr> traderOrders;
    for (const auto& order : activeOrders) {
        if (order->getTraderId() == traderId) {
            traderOrders.append(order);
        }
    }
    return traderOrders;
}

QMap<QString, QVariant> TradingEngine::getSymbolStatistics(const QString& symbol) const {
    QMap<QString, QVariant> stats;
    
    if (!orderBooks.contains(symbol)) {
        return stats;
    }
    
    auto orderBook = orderBooks[symbol];
    auto recentTrades = getRecentTradesForSymbol(symbol, 100);
    
    // Calculate price statistics
    if (!recentTrades.isEmpty()) {
        QVector<double> prices;
        for (const auto& trade : recentTrades) {
            prices.append(trade.price);
        }
        
        std::sort(prices.begin(), prices.end());
        
        double highPrice = prices.last();
        double lowPrice = prices.first();
        double lastPrice = recentTrades.last().price;
        
        // Calculate VWAP
        double totalValue = 0.0;
        int totalVolume = 0;
        for (const auto& trade : recentTrades) {
            totalValue += trade.price * trade.quantity;
            totalVolume += trade.quantity;
        }
        double vwap = totalVolume > 0 ? totalValue / totalVolume : 0.0;
        
        stats["symbol"] = symbol;
        stats["last_price"] = lastPrice;
        stats["high_price"] = highPrice;
        stats["low_price"] = lowPrice;
        stats["vwap"] = vwap;
        stats["total_volume"] = totalVolume;
        stats["trade_count"] = recentTrades.size();
    }
    
    return stats;
}

void TradingEngine::clear() {
    QMutexLocker orderLocker(&orderQueueMutex);
    QMutexLocker tradeLocker(&tradeHistoryMutex);
    QMutexLocker statsLocker(&statsMutex);
    
    orderBooks.clear();
    activeOrders.clear();
    orderQueue.clear();
    tradeHistory.clear();
    latencyMeasurements.clear();
    
    totalTrades = 0;
    totalVolume = 0;
    ordersPerSecond = 0;
    ordersProcessedSinceLastUpdate = 0;
}

void TradingEngine::processOrders() {
    if (!running) {
        return;
    }
    
    // Process up to 10 orders per cycle to prevent blocking
    int ordersToProcess = qMin(10, orderQueue.size());
    
    for (int i = 0; i < ordersToProcess; ++i) {
        QMutexLocker locker(&orderQueueMutex);
        
        if (orderQueue.isEmpty()) {
            break;
        }
        
        OrderPtr order = orderQueue.dequeue();
        locker.unlock();
        
        processOrder(order);
    }
}

void TradingEngine::updateStats() {
    QMutexLocker locker(&statsMutex);
    
    qint64 elapsed = lastStatsUpdate.restart();
    if (elapsed > 0) {
        ordersPerSecond = static_cast<int>((ordersProcessedSinceLastUpdate * 1000.0) / elapsed);
        ordersProcessedSinceLastUpdate = 0;
    }
    
    // Emit performance stats update
    emit performanceStatsUpdated(getPerformanceStats());
}

void TradingEngine::processOrder(OrderPtr order) {
    QElapsedTimer processTimer;
    processTimer.start();
    
    QMutexLocker locker(&orderQueueMutex);
    
    // Add to active orders
    activeOrders[order->getOrderId()] = order;
    
    // Get order book
    auto orderBook = getOrderBook(order->getSymbol());
    
    locker.unlock();
    
    // Try to match the order
    matchOrder(order, orderBook);
    
    locker.relock();
    
    // If order still has quantity, add to book
    if (order->isActive() && order->getQuantity() > 0) {
        orderBook->addOrder(order);
    } else {
        // Remove from active orders if completely filled or cancelled
        if (activeOrders.contains(order->getOrderId())) {
            activeOrders.remove(order->getOrderId());
        }
    }
    
    // Record processing latency
    qint64 latencyMs = processTimer.elapsed();
    
    QMutexLocker statsLocker(&statsMutex);
    latencyMeasurements.append(latencyMs);
    
    // Keep only recent measurements
    if (latencyMeasurements.size() > MAX_LATENCY_MEASUREMENTS) {
        latencyMeasurements.removeFirst();
    }
    
    ordersProcessedSinceLastUpdate++;
}

void TradingEngine::matchOrder(OrderPtr incomingOrder, std::shared_ptr<OrderBook> orderBook) {
    if (incomingOrder->getSide() == OrderSide::BUY) {
        matchBuyOrder(incomingOrder, orderBook);
    } else {
        matchSellOrder(incomingOrder, orderBook);
    }
}

void TradingEngine::matchBuyOrder(OrderPtr buyOrder, std::shared_ptr<OrderBook> orderBook) {
    while (buyOrder->getQuantity() > 0 && buyOrder->isActive()) {
        OrderPtr bestAsk = orderBook->getBestAsk();
        
        if (!bestAsk || bestAsk->getPrice() > buyOrder->getPrice()) {
            break; // No more matching orders
        }
        
        // Execute trade
        int tradeQuantity = qMin(buyOrder->getQuantity(), bestAsk->getQuantity());
        double tradePrice = bestAsk->getPrice(); // Price-time priority: use ask price
        
        executeTrade(buyOrder, bestAsk, tradeQuantity, tradePrice, orderBook);
        
        // Remove ask order if completely filled
        if (bestAsk->getQuantity() == 0) {
            orderBook->removeOrder(bestAsk->getOrderId(), OrderSide::SELL);
            
            QMutexLocker locker(&orderQueueMutex);
            if (activeOrders.contains(bestAsk->getOrderId())) {
                activeOrders.remove(bestAsk->getOrderId());
            }
        }
    }
}

void TradingEngine::matchSellOrder(OrderPtr sellOrder, std::shared_ptr<OrderBook> orderBook) {
    while (sellOrder->getQuantity() > 0 && sellOrder->isActive()) {
        OrderPtr bestBid = orderBook->getBestBid();
        
        if (!bestBid || bestBid->getPrice() < sellOrder->getPrice()) {
            break; // No more matching orders
        }
        
        // Execute trade
        int tradeQuantity = qMin(sellOrder->getQuantity(), bestBid->getQuantity());
        double tradePrice = bestBid->getPrice(); // Price-time priority: use bid price
        
        executeTrade(bestBid, sellOrder, tradeQuantity, tradePrice, orderBook);
        
        // Remove bid order if completely filled
        if (bestBid->getQuantity() == 0) {
            orderBook->removeOrder(bestBid->getOrderId(), OrderSide::BUY);
            
            QMutexLocker locker(&orderQueueMutex);
            if (activeOrders.contains(bestBid->getOrderId())) {
                activeOrders.remove(bestBid->getOrderId());
            }
        }
    }
}

void TradingEngine::executeTrade(OrderPtr buyOrder, OrderPtr sellOrder, 
                                int quantity, double price, std::shared_ptr<OrderBook> orderBook) {
    // Fill both orders
    buyOrder->fill(quantity, price);
    sellOrder->fill(quantity, price);
    
    // Create trade record
    Trade trade(generateTradeId(), buyOrder->getSymbol(), quantity, price,
                buyOrder->getTraderId(), sellOrder->getTraderId(),
                buyOrder->getOrderId(), sellOrder->getOrderId());
    
    // Add to trade history
    {
        QMutexLocker locker(&tradeHistoryMutex);
        tradeHistory.append(trade);
        
        // Keep only recent trades
        if (tradeHistory.size() > MAX_TRADE_HISTORY) {
            tradeHistory.removeFirst();
        }
        
        totalTrades++;
        totalVolume += quantity;
    }
    
    // Add to order book trade history
    orderBook->addTrade(trade);
    
    // Emit signals
    emit tradeExecuted(trade);
    emit orderFilled(buyOrder, quantity, price);
    emit orderFilled(sellOrder, quantity, price);
    
    // Notify traders of fills
    notifyTraderFill(buyOrder, quantity, price);
    notifyTraderFill(sellOrder, quantity, price);
}

void TradingEngine::notifyTraderFill(OrderPtr order, int quantity, double price) {
    QString traderId = order->getTraderId();
    if (traders.contains(traderId)) {
        try {
            Trader* trader = traders[traderId];
            trader->onOrderFilled(order, quantity, price);
        } catch (...) {
            qWarning() << "Error notifying trader" << traderId;
        }
    }
}

QString TradingEngine::generateTradeId() {
    return QString("%1").arg(totalTrades + 1, 6, 10, QChar('0'));
}